doc/source/cluster/node-provider.rst

```diff
@@ -0,0 +1,142 @@
+#. External_ips are used in "Ray up" when starting the head node. Internal_ips are used for communication within the cluster.
+#. Nodes should appear immediately once started by `create_node`, and transition immediately to terminated when `terminate_node` is called. This means any APIs, that are affected by the nodes states, especially non_terminate_nodes(), should reflect the changes immediately.
+#. Node tags should be readable by node_tags() immediately after set_node_tags() returns.  
+#. The instance of this class can be constructed and destructed by multiple threads simultaneously. As a result:
+
+    1. All data operations should be thread-safe
+    2. State information should not be stored as class variables that are lost when the class instance is destructed
+
+
+Autoscaler Config YAML
+------------------------
+
+A customized NodeProvider goes with a customized autoscaler config YAML. Besides the parts that is required by the launching process and the autoscaler, some parts of the autoscaler config YAML is only used by the NodeProvider. As a result, those parts can also have any kind of format that helps the implementation of NodeProvider.
+
+The YAML file section that is only used by NodeProvider:
+
+* provider (except for "type" and "module")
+* available_node_types.*.node_config
+* auth (for command runner)
+
+Autoscaler tries to fill the fields available in the "minimal default config" (see ray.autoscaler._private.providers._get_default_config for detail), if those fields are not provided by user. After the config is filled, the NodeProvider can further modify it during the launching process, using the cluster config APIs (the first API category). 
+
+Advance
+------------------------
+
+A general overview of how autoscaler is connected to node_provider:
+
+.. image:: node-provider.png
+   :width: 600
+
+#. The booting process from command line includes: (see autoscaler._private.command)
+
+    1. Filling the autoscaler config 
+    2. Starting the head node
+    3. Copying the modified autoscaling config to the head node. 
+
+#. All the worker nodes are created by the autoscaler, a daemon running at the head node, which is not a part of the booting process. (See autoscaler._private.autoscaler)
+#. The autoscaler starts a new thread (node launcher) for each batches of node creation. (see autoscaler._private.node_launcher)
+#. The autoscaler starts another thread (node updater, which is different from the node creation thread) to set up the nodes. If the node is not ready to run a command, this thread keeps trying until success or raises failure after timeout. (See autoscaler._private.updater)
+#. The node termination is done by the main autoscaler thread, based on the node statistic collected from GCS or the failure raises by node updater. 
+
+Possible Hacking
```

Comment:
While `Hacking` does accurately reflect the current state of things re: [updaters]/[no updaters], we'll want to soften the language a bit for the docs :)

Also, there's a hidden flag to disable node updaters that we could document here.